// permissionless/_cjs/utils/encodeInstallModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeInstallModule = encodeInstallModule;
const viem_1 = require("viem");
const supportsModule_1 = require("../actions/erc7579/supportsModule.js");
const errors_1 = require("../errors/index.js");
function encodeInstallModule(parameters) {
    const account = parameters.account;
    if (!account) {
        throw new errors_1.AccountNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const modules = Array.isArray(parameters.modules)
        ? parameters.modules
        : [parameters.modules];
    return modules.map(({ type, address, context, initData }) => ({
        to: account.address,
        value: BigInt(0),
        data: (0, viem_1.encodeFunctionData)({
            abi: [
                {
                    name: "installModule",
                    type: "function",
                    stateMutability: "nonpayable",
                    inputs: [
                        {
                            type: "uint256",
                            name: "moduleTypeId"
                        },
                        {
                            type: "address",
                            name: "module"
                        },
                        {
                            type: "bytes",
                            name: "initData"
                        }
                    ],
                    outputs: []
                }
            ],
            functionName: "installModule",
            args: [
                (0, supportsModule_1.parseModuleTypeId)(type),
                (0, viem_1.getAddress)(address),
                context ?? initData
            ]
        })
    }));
}
